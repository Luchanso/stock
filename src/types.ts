export enum Currency {
  usd = 'USD',
  rub = 'RUB'
}

export enum ActionTypes {
  add,
  remove,
  fee
}

export type StockShareLot = {
  count: number;
  price: number;
  currency: Currency;
  symbol: string;
};

export type BasicAction = {
  lot: StockShareLot;
  time: number;
  description?: string;
};

export type AddStockAction = {
  type: ActionTypes.add;
} & BasicAction;

export type RemoveStockAction = {
  type: ActionTypes.remove;
} & BasicAction;

export type FeeAction = {
  type: ActionTypes.fee;
  amount: number;
  currency: Currency;
  description?: string;
};

export type Action = AddStockAction | RemoveStockAction | FeeAction;
