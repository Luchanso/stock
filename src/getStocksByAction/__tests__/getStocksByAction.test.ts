import { Action, ActionTypes, Currency } from "../../types";
import { getStocksByAction } from "../index";

test("should return correct stock list", () => {
  const result = {
    AAA: 120,
    BBB: 4
  };
  const defaultActionList: Action[] = [
    {
      type: ActionTypes.add,
      lot: {
        count: 123,
        currency: Currency.usd,
        price: 10.999,
        symbol: "AAA"
      },
      time: new Date(2018).getTime() / 1000
    },
    {
      type: ActionTypes.add,
      lot: {
        count: 3,
        currency: Currency.rub,
        price: 3,
        symbol: "BBB"
      },
      time: new Date(2018).getTime() / 1000
    },
    {
      type: ActionTypes.remove,
      lot: {
        count: 3,
        currency: Currency.usd,
        price: 19,
        symbol: "AAA"
      },
      time: new Date(2019).getTime() / 1000
    },
    {
      type: ActionTypes.fee,
      amount: 10,
      currency: Currency.usd
    },
    {
      type: ActionTypes.add,
      lot: {
        count: 1,
        currency: Currency.rub,
        price: 2,
        symbol: "BBB"
      },
      time: new Date(2019).getTime() / 1000
    }
  ];

  expect(getStocksByAction(defaultActionList)).toStrictEqual(result);
});

test("Should return error when stock history incorrect", () => {
  const incorectActionList: Action[] = [
    {
      type: ActionTypes.remove,
      lot: {
        count: 123,
        currency: Currency.rub,
        price: 1,
        symbol: "AAA"
      },
      time: new Date(2018).getTime() / 1000
    }
  ];

  expect(
    () => { getStocksByAction(incorectActionList) }
  ).toThrowErrorMatchingSnapshot();
});
