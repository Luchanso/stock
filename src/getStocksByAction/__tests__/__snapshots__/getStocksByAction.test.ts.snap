// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return error when stock history incorrect 1`] = `
"Not found stock in history {
  \\"type\\": 1,
  \\"lot\\": {
    \\"count\\": 123,
    \\"currency\\": 1,
    \\"price\\": 1,
    \\"symbol\\": \\"AAA\\"
  },
  \\"time\\": 2.018
}"
`;
